smart-contracts/test/Unlock/swapAndCall.mainnet.js

```diff
@@ -0,0 +1,262 @@
+const { ethers, upgrades } = require('hardhat')
+const { expect } = require('chai')
+const { 
+  impersonate, 
+  getUniswapTokens, 
+  getUniswapRoute,
+  ADDRESS_ZERO, 
+  deployLock, 
+  purchaseKey,
+  getBalance,
+  WETH,
+  addERC20,
+  UNLOCK_PROXY_OWNER,
+  UNLOCK_ADDRESS,
+  PERMIT2_ADDRESS,
+  CHAIN_ID
+ } = require('../helpers')
+
+
+// get uniswap-formatted tokens
+const tokens = getUniswapTokens(CHAIN_ID)
+
+const scenarios = [
+  [tokens.native, tokens.dai],
+  [tokens.native, tokens.usdc],
+  [tokens.dai, tokens.usdc],
+  [tokens.usdc, tokens.native],
+  [tokens.udt, tokens.native],
+  [tokens.native, tokens.wBtc]
+]
+
+describe(`swapAndCall`, function() {
+
+  let unlock
+  before(async function() {
+    if (!process.env.RUN_FORK) {
+      // all suite will be skipped
+      this.skip()
+    }
+
+    // get Unlock contract
+    unlock = await ethers.getContractAt('Unlock', UNLOCK_ADDRESS)
+    
+    // upgrade Unlock with the modified version
+    await impersonate(UNLOCK_PROXY_OWNER)
+    const unlockProxyOwner = await ethers.getSigner(UNLOCK_PROXY_OWNER)
+    const Unlock = await ethers.getContractFactory('Unlock', unlockProxyOwner)
+    await upgrades.upgradeProxy(UNLOCK_ADDRESS, Unlock)
+
+    // get unlock owner
+    const unlockOwnerAddress = await unlock.owner()
+    await impersonate(unlockOwnerAddress)
+    const unlockOwner = await ethers.getSigner(unlockOwnerAddress)
+    
+    // config unlock
+    await unlock.connect(unlockOwner).configUnlock(
+      ADDRESS_ZERO, // udt
+      WETH,
+      16000, // gasEstimate
+      'KEY_SWAP',
+      'http://locksmith:8080/api/key/',
+      CHAIN_ID, // fork
+    )
+
+    // config permit2
+    await unlock.connect(unlockOwner).setPermit2(PERMIT2_ADDRESS);
+  })
+
+  it('permit2 is set properly', async () => {
+    expect(await unlock.permit2()).to.equal(PERMIT2_ADDRESS)
+  })
+
+  scenarios.forEach(([srcToken, lockToken]) => {
+    let keyOwner, lock, keyPrice
+    let swapCalldata, value, swapRouter, amountInMax
+    describe(`use ${srcToken.symbol} with a lock priced in ${lockToken.symbol}`, () => {
+      before(async () => {
+        ;[, keyOwner] = await ethers.getSigners()
+        // parse token decimals properly (100 USDC or 1 ETH)
+        keyPrice = ethers.utils.parseUnits(
+          lockToken.symbol === 'USDC' ? '100' : '1', 
+          lockToken.decimals
+        )
+        lock = await deployLock({
+          unlock,
+          tokenAddress: lockToken.address,
+          // make sure we can add multiple keys
+          maxKeysPerAddress: 100,
+          keyPrice,
+          isEthers: true
```

Comment:
What's this?

Reply:
We still use truffle in most tests so when deploying the contracts with the helpers it will return a truffle object. That's just a switch to parse the contract as an Etherjs instance